Fix exception on FeatureGroup getBounds() with LayerGroup/FeatureGroup as child

rafaelgalani opened the pull request:
Added logic to handle groups (LayerGroup/FeatureGroup) as children of a FeatureGroup. It will also work with nested groups.
Unit tests were also added to check the new logic. There was no need to modify the documentation, since the current one already expected such behaviour.

This PR fixes #7228.

Falke-Design commented:
Nice PR, maybe you can also add something what break the loop if a Group has already called `getBounds()`.

```
var fg = L.featureGroup();
var lg = L.featureGroup().addTo(fg);
fg.addTo(lg);
console.log(fg.getBounds());
```

This ends in a infinity loop, because both FeatureGroups has the other FeatureGroup as child

rafaelgalani commented:
@Falke-Design Yup, it loops forever. I wasn't expecting this scenario - to be honest, I didn't even thought that someone could do this. I'll add it to the unit tests and fix it. Thanks!

rafaelgalani commented:
@Falke-Design again, thanks for the feedback. The issue you pointed was fixed, I guess it's all OK now.

rafaelgalani commented:
Oops. 
Accidentally closed the pull request. Just reopened it now... Sorry about it 😅

rafaelgalani commented:
@IvanSanchez I'm not sure whether you saw the PR, but it's been some time since I opened it.. Since it hasn't been assigned a reviewer, I thought of mentioning you. 
Please let me know if you think that there's anything that could be improved 😸

johnd0e commented:
And question: do we have recursive processing for `getBounds` only?

Falke-Design commented:
@rafaelgalani can you please resolve the suggestions

Falke-Design commented:
> And question: do we have recursive processing for getBounds only?

Same here: #7273

rafaelgalani commented:
Sorry y'all for leaving this one behind. I'll rebase on `main` and work on it asap.

rafaelgalani commented:
I really dislike what I did on this one. I don't really know what I had in mind 😬 
I updated and refactored most of it, and took a different approach to handling recursion:

I think we should prevent it (as added by @johnd0e) instead of adding workarounds - also, this was a very hacky and _suboptimal_ workaround.

Please let me know if that looks better.

Falke-Design commented on src/layer/LayerGroup.js:
```diff
@@ -50,6 +52,26 @@ export const LayerGroup = Layer.extend({
+	hasDescendant(layer) {
+		let hasDescendant = false;
+
+		this.eachLayer((childLayer) => {
+			if (childLayer === layer || (childLayer instanceof LayerGroup && childLayer.hasDescendant(layer))) {
+				hasDescendant = true;
+			}
+		});
```
Changing this to a for-loop would make it possible to break it if a descendant has been found.

Falke-Design commented on src/layer/LayerGroup.js:
```diff
@@ -50,6 +52,26 @@ export const LayerGroup = Layer.extend({
 		return this;
 	},
 
+	_preventLayerGroupRecursion(layer) {
+		if (layer instanceof LayerGroup && layer.hasDescendant(this)) {
+			throw new Error('Cannot add a LayerGroup as a child if it is already included, directly or nested.');
+		}
+	},
```
This is a huge breaking-change, I don't know what I should thing about it

rafaelgalani replied:
Ummm, you're right. Sorry, I didn't think about this specifically...

So, in this case, I think we'd want to go with the recursive solution. It would be a non-breaking change, and it would also solve both issues.

rafaelgalani commented:
@Falke-Design Howdy! 
I updated the PR to go with the recursive behavior. LMK what you think. 

I still would like to understand a little more about the recursive use cases, but that's for another PR/issue 😬

Falke-Design commented on src/layer/LayerGroup.js:
```diff
@@ -147,6 +148,23 @@ export const LayerGroup = Layer.extend({
+			}
+
+			visitedIds.add(layerId);
+			bounds.extend(layer instanceof LayerGroup ? layer.getBounds(visitedIds) : layer.getLatLng());
```
Not all layers have `.getLatLng()`, so we need to check for `getBounds()` (ImageOverlay) and maybe `getLatLngs()` too. I would add also a check for `getLatLng`, because maybe a layer is included with no coordinates.

```suggestion
if(layer instanceof LayerGroup || layer.getBounds){
			bounds.extend(layer.getBounds(visitedIds));
} else if(layer.getLatLngs){
			bounds.extend(layer.getLatLngs());
} else if(layer.getLatLng){
			bounds.extend(layer.getLatLng());
}
```
bounds.extend(layer.getBounds ? layer.getBounds() : layer.getLatLng());

Falke-Design commented on src/layer/LayerGroup.js:
```diff
@@ -147,6 +148,23 @@ export const LayerGroup = Layer.extend({
+
+	// @method getBounds(): LatLngBounds
+	// Returns the LatLngBounds of the Layer Group (created from bounds and coordinates of its children).
+	getBounds(visitedIds = new Set()) {
```
Let us make the variable `visitedIds` "private":
```suggestion
	getBounds(_visitedIds = new Set()) {

```

Falke-Design reviewed (changes requested):
You included the recursive logic for `getBounds()` but no test.

Do something like this:
```
const childGroup = new FeatureGroup();
const parentGroup = new FeatureGroup([childGroup]);
childGroup.addLayer(parentGroup);

parentGroup.getBounds(); --> No error
```

--------------

I still would like to get this recursive solution but I don't know if it is the best. Maybe we should prevent it like your solution but this will be a breaking change. 
I think for now we should go with the recursive solution, but maybe another maintainer can let us know what he thinks

rafaelgalani commented:
@Falke-Design Thanks for the review! I just updated the PR with your suggestions, lmk what you think.

Falke-Design commented on spec/suites/layer/LayerGroupSpec.js:
```diff
@@ -112,4 +112,64 @@
+			expect(fg.getBounds()).to.eql(bounds);
+		});
+
+		it('doesn\'t break if there are nested layer groups and returns the bounds of the nested groups', () => {
+			const southWest = L.latLng(0, 0),
+			northEast = L.latLng(4, 4);
+
+			const childGroup = L.featureGroup([
+				L.marker(southWest),
+				L.marker(northEast),
+			]);
+
+			const parentGroup = L.layerGroup([
+				childGroup
+			]);
+
+			const bounds = L.latLngBounds(southWest, northEast);
+
+			let result = [];
+			const callback = () => {
+				result = parentGroup.getBounds();
+			};
+			expect(callback).to.not.throwError();
+			expect(result).to.eql(bounds);
+		});
+	});
```
This works with the current solution too, you need to add to the `childGroup` the `parenGroup`: `childGroup.addLayer(parentGroup)`

rafaelgalani replied:
Oops, forgot to add it. My bad.

rafaelgalani commented:
☝🏻 Sorry for that, clicked it by accident

rafaelgalani commented:
Rebase on `main` ☝🏻

jonkoops commented on src/layer/LayerGroup.js:
```diff
@@ -147,6 +148,29 @@ export const LayerGroup = Layer.extend({
+			}
+
+			_visitedIds.add(layerId);
+			if (layer instanceof LayerGroup || layer.getBounds) {
```
I think a `typeof` check would suffice here. Just thinking out loud if all entities passed here have some kind of base class with a `getBounds()` method (in which case we could `instanceof` that base class).

Falke-Design replied:
Let us start with a typeof check
```suggestion
			if (layer instanceof LayerGroup || typeof layer.getBounds === 'function') {
```
